### PR TITLE
Kernel

### PR DESCRIPTION
**Pull request by virgile-baudrot**
# Description

Additional kernels: Geometric, Gaussian and Bivariate Student 
Formulation founds in Nathan 2012, "Dispersal kernels: review" from the book Dispersal Ecology and Evolution (2012) and Walker et al. (2019) "A Spatio-Temporal Exposure-Hazard Model for Assessing Biological Risk and Impact"

❓  How (where) to cite the sources?

## Changes (choose one only, with x or click after creating PR):

- [x] New formulation/rule
- [ ] Documentation
- [ ] Bug fix for existing rules (non-breaking syntax)
- [ ] Minor change to existing rules (no breaks to old syntax)
- [ ] Major change to existing rules (breaking syntax, needs a breaking version bump)

# Checklist:

- [x] My code follows the [BlueStyle](https://github.com/invenia/BlueStyle) style guide
- [x] My code is commented with explanations of anything hard to understand
- [x] There are no unnecessary or unrelated code changes included in this PR
- [x] There are no dependencies added: if so, please consider if they are necessary and explain why they are required

### For bugfixes

- [ ] My PR addresses a single bug/connected group of bugs
- [ ] I have included a test that will ensure the bug does not reoccur

### For new `Rule`s

- [x] My PR is limited to a clear theme/conceptual unit (otherwise make multiple PRs that can be discussed separately).
- [x] New rules have tests in a file of the same name as the rule file, e.g. `/test/newrulename.jl`
- [ ] I have cited the source of the new formulation/s, if applicable
- [x] I have added a documentation string to formulation struct/s, and added an entry in the docs

### CODE REVIEW

**Review by jamesmaino — APPROVED** [review]
Great! Perhaps we should cite in the documentation string for each function?

File: src/Dispersal.jl [hunk 1]
```diff
@@ -20,7 +20,7 @@ using ModelParameters: params
 
 import DynamicGrids: applyrule, applyrule!, precalcrule, kernel
 
-export KernelFormulation, ExponentialKernel
+export KernelFormulation, ExponentialKernel, GeometricKernel, GaussianKernel, BivariateStudentKernel
```
Comment (rafaqz) [inline src/Dispersal.jl]: Do you think we should remove `Kernel` from the end of all of these? I would be cleaner for scripting, and its a bit confusing that `DisperselKernel` also ends in `Kernel`.

File: src/kernel/formulations.jl [hunk 2]
```diff
@@ -32,3 +32,69 @@ where λ is a shape parameter.
...
+```math
+y = \frac{(α+1)(α+2)}{2 π} (1+d)^α
+```
+
+where α is a shape parameter.
+"""
+@Base.kwdef struct GeometricKernel{P} <: KernelFormulation
+    α::P = Param(1.0, bounds=(-1000.0, 1000.0))
+end
+(f::GeometricKernel)(distance) = (1+distance)^f.α * ((f.α+1)*(f.α+2))/(2*π)
```
Comment (rafaqz) [inline src/kernel/formulations.jl]: spaces around math, except `^`

And `d`  for `distance`?

File: src/kernel/formulations.jl [hunk 3]
```diff
@@ -32,3 +32,69 @@ where λ is a shape parameter.
     λ::P = Param(1.0, bounds=(0.0, 2.0))
 end
 (f::ExponentialKernel)(distance) = exp(-distance / f.λ)
```
Comment (rafaqz) [inline src/kernel/formulations.jl]: Should we use `d` everywhere for distance, seeing it is used below?

File: src/kernel/formulations.jl [hunk 4]
```diff
@@ -32,3 +32,69 @@ where λ is a shape parameter.
...
+```math
+y = \frac{(α+1)(α+2)}{2 π} (1+d)^α
+```
+
+where α is a positive parameter.
+"""
+@Base.kwdef struct GaussianKernel{P} <: KernelFormulation
+    α::P = Param(1.0, bounds=(0.0, 1000.0))
+end
+(f::GaussianKernel)(d) = 1/(π*f.α^2)*exp(-d^2/f.α^2)
```
Comment (rafaqz) [inline src/kernel/formulations.jl]: spaces around math except `^` and singular `-`

File: src/kernel/formulations.jl [hunk 5]
```diff
@@ -32,3 +32,69 @@ where λ is a shape parameter.
...
+```
+"""
+@Base.kwdef struct BivariateStudentKernel{P1,P2,P3,P4,P5} <: KernelFormulation
+    α::P1 = Param(1.0, bounds=(0.0, 1000.0))
+    β::P2 = Param(0.5, bounds=(0.0, 1.0))
+    θ::P3 = Param(0.5, bounds=(0.0, 10.0))
+    c1::P4 = Param(0.5, bounds=(0.0, 10.0))
+    c2::P5 = Param(0.5, bounds=(0.0, 2*π))
+end
+(f::BivariateStudentKernel)(d) = (f.β-1)/(π*f.α^2)*(1+d^2/f.α^2)*exp(f.c1 * cos(f.θ-f.c2))
```
Comment (rafaqz) [inline src/kernel/formulations.jl]: spaces around math

File: test/kernels.jl [hunk 6]
```diff
@@ -165,3 +166,29 @@ end
     end
 
 end
+
+@testset "kernel formulation" begin
+    @testset "ExponentialKernel" begin
+        hood = DispersalKernel(; formulation=ExponentialKernel(0.5))
+        @test DynamicGrids.radius(hood) == 1
+        @test sum(DynamicGrids.kernel(hood)) ≈ 1 atol=10^(-10)
```
Comment (rafaqz) [inline test/kernels.jl]: `1e-10` for `10^(-10)`, and the rest below

File: src/kernel/formulations.jl [hunk 7]
```diff
@@ -32,3 +32,69 @@ where λ is a shape parameter.
...
+(f::GaussianKernel)(d) = 1/(π*f.α^2)*exp(-d^2/f.α^2)
+
+
+"""
+    BivariateStudentKernel <: KernelFormulation
+
+    BivariateStudentKernel(α, β, θ, c1, c2)
+
+Probability density function of distance ``d``.
+The Bivariate Student kernel has a power-law decrease and is anisotropic.
```
Comment (rafaqz) [inline src/kernel/formulations.jl]: Maybe explain it in slightly more lay terms as well, maybe give an example use-case for a directional kernel?

**Review by rafaqz — CHANGES_REQUESTED** [review]
This is great. Thanks for writing these up.

I'm wondering if we should drop the `Kernel` from the end, so you just do `formulation=Gaussian(a)`. It would be nicer for scripts.

I would put the citation in every doc string.

**Comment by virgile-baudrot** [comment]
>  drop the `Kernel` from the end, so you just do `formulation=Gaussian(a)`

I thought about that also. For Gaussian and Geometric it's working. But we have `ExponentialKernel`, `ExponentialGrowth` and `ExponentialMortality`... what makes many exponential, and why calling one Exponential and not the other

File: src/kernel/formulations.jl (hunk 7 above)
Comment (virgile-baudrot) [reply src/kernel/formulations.jl]: ok. and the parameterisation I used is isotropic 👎

File: src/kernel/formulations.jl (hunk 3 above)
Comment (virgile-baudrot) [reply src/kernel/formulations.jl]: Yes `d` is fine. `distance` is quite big when formulation is long. There is also `r` as radius which is often used (in Nathan 2012).

**Comment by rafaqz** [comment]
That's a good point. `Exponential` needs to be clarified. `Geometric` is also the same name as `Distributions.Geometric` - so maybe leaving `Kernel` on the end is better.

**Comment by rafaqz** [comment]
Any movement on this one virgile? Youve probably been busy. But great to have it in the package for submission

File: src/Dispersal.jl (hunk 1 above)
Comment (virgile-baudrot) [reply src/Dispersal.jl]: We have other rules using word like Exponential in `ExponentialGrowth` for instance, so keeping `Kernel` key word is maybe valuable.

**Comment by virgile-baudrot** [comment]
I made some updates on type of wirtting equations.
Also I added Weibull kernel which is very use for pollen dispersal
I removed BivarianteStudent, I'm not fully sure about the mathematical formulation in Nathan et al.

File: src/kernel/formulations.jl [hunk 8]
```diff
@@ -31,4 +31,63 @@ where λ is a shape parameter.
...
+
+The Geometric kernel has a power-law decrease and is isotropic.
+
+```math
+y = \frac{(α+1)(α+2)}{2 π} (1+d)^α
+```
+
+where α is a shape parameter.
+"""
+@Base.kwdef struct GeometricKernel{P} <: KernelFormulation
```
Comment (rafaqz) [inline src/kernel/formulations.jl]: This `@` should be before `kwdef` :)

File: src/kernel/formulations.jl [hunk 9]
```diff
@@ -31,4 +31,63 @@ where λ is a shape parameter.
...
+
+Probability density function of distance ``d``.
+
+```math
+y = \frac{1}{π α^2} \exp\left( \frac{-d^2}{α^2}  \right) 
+```
+
+where α is a positive parameter.
+"""
+@Base.kwdef struct GaussianKernel{P} <: KernelFormulation
```
Comment (rafaqz) [inline src/kernel/formulations.jl]: This one too

File: src/kernel/formulations.jl [hunk 10]
```diff
@@ -31,4 +31,63 @@ where λ is a shape parameter.
...
+
+Probability density function of distance ``d``.
+
+```math
+y = \frac{β}{2 π α^2} d^{β-2} \exp\left( \frac{-d^β}{α^β}  \right) 
+```
+
+where α and β are positive parameters.
+"""
+@Base.kwdef struct WeibullKernel{A,B} <: KernelFormulation
```
Comment (rafaqz) [inline src/kernel/formulations.jl]: And this

**Comment by rafaqz** [comment]
Also your commits are a bit messed up with some of my commits somehow, and your mortality tests. Might need to start a new branch from master, recommit the kernel files, and push here to this branch again.

When working on a PR you should always `rebase` it against master and other branches, instead of `merge`. `rebase` just rewrites all your commits at the end.

**Comment by virgile-baudrot** [comment]
arrff, sorry for that. I made a change on my forks forgetting it follows the PR here

I'm doing a new branch base on the master for kernel... come in few minutes

**Comment by virgile-baudrot** [comment]
@rafaqz how do I PR here with a this new branch [kernel_PR](https://github.com/virgile-baudrot/Dispersal.jl/commit/6197a0c214edf3547f1f8561e5e3967d96690cde) which is on my fork? Should I start a new PR and close this one?

**Comment by rafaqz** [comment]
You can push another branch to this one with `--force`

Or start a new PR, up to you.

**Comment by virgile-baudrot** [comment]
move to #98

**Pull request closed without merging**